webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/campaigns/MainCampaignService.scala

```diff
@@ -92,6 +104,16 @@ class MainCampaignService(repo: CampaignEventRepository, campaignRepo: CampaignR
 
   case class CampaignScheduler(main: MainCampaignService, queue: Queue[CampaignEventId], zclock: Clock) {
 
+    def deleteCampaign(c : CampaignId) = {
+      for {
+        campaign <- campaignRepo.get(c)
+        events <- repo.getWithCriteria(campaignId = Some(c))
+        _ <- ZIO.foreach(events) { event => ZIO.foreach(services)(s => s.delete(main, event)(campaign)).catchAll(_ => CampaignLogger.warn(s"An error occured while cleaning campaign event ${event.id.value} during deletion of campaign ${c.value}") ) }
```

Comment:
I'm not sure what are the consequences here. So we are in the case where we try to delete campaign event, the campaign is already deleted. Is there a way the user can correct the situation?

Reply:
The campaign is not deleted is there is an error here, you can still delete it again

the workflow is 

User decides to delete a campaign
main campaign service delegates to it inner services handler (l81-83)
inner services handler,
=> get the campaign (L 109
=> Get events linked to the campaign (L 110
=> Clean events if needed (line 111) (like  if they were running state )
=> delete events (L 112)
going back to parents, which deletes the campaign (L 84)


So if anything goes wrong you can run it again, until all events are cleaned, then deleted, then campaign is deleted